app/lib/lti_advantage/user.rb

```diff
@@ -0,0 +1,126 @@
+module LtiAdvantage
+  class User
+    def initialize(lti_token, application_instance)
+      @lti_token = lti_token
+      @application_instance = application_instance
+      @lti_user_id = lti_token["sub"]
+    end
+
+    def user
+      user = ::User.find_by(lti_user_id: @lti_user_id)
+      if user.blank?
+        user = _generate_new_lti_user
+        _attempt_uniq_email(user)
+      else
+        if user.lms_user_id.blank? && _lms_user_id.present?
+          user.lms_user_id = _lms_user_id
+        end
+        _update_roles(user)
+      end
+      user
+    end
+
+    def name
+      return "anonymous" if @application_instance.anonymous?
+
+      @lti_token["name"] ||
+        "#{@lti_token['given_name']} #{@lti_token['family_name']}" ||
+        @lti_token["email"]
+    end
+
+    def email
+      if @application_instance.anonymous?
+        _generate_email
+      else
+        # If there isn't an email then we have to make one up. We use the user_id and instance guid
+        email = @lti_token["email"]
+        email = "user-#{@lti_user_id}@#{_domain_for_email}" if email.blank? && @lti_user_id.present?
+        # If there isn't an email then we have to make one up. We use the user_id and instance guid
+        email = "generated-#{lti_user_id}@#{_domain_for_email}" if email.blank?
```

Comment:
Why do we have both line 37 and line 39? I see that they're doing slightly different things, but I'm not sure I understand why they're both necessary since we're making up an email either way.

Reply:
Good catch. I don't think we need both. I'll clean that up.